File: common/src/main/java/net/impleri/playerskills/integration/kubejs/api/AbstractRegistrationEventJS.java
package net.impleri.playerskills.integration.kubejs.api;

import dev.latvian.mods.kubejs.server.ServerEventJS;
import dev.latvian.mods.kubejs.util.ConsoleJS;
import dev.latvian.mods.rhino.util.HideFromJS;
import net.impleri.playerskills.restrictions.AbstractRestriction;
import net.impleri.playerskills.utils.RegistrationType;
import net.minecraft.core.Registry;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.MinecraftServer;
import net.minecraft.tags.TagKey;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

abstract public class AbstractRegistrationEventJS<T, R extends AbstractRestriction<T>, B extends AbstractRestrictionBuilder<R>> extends ServerEventJS {
    private final ResourceKey<Registry<T>> registryName;

    private final Registry<T> registry;

    private final String type;

    protected final MinecraftServer server;

    public AbstractRegistrationEventJS(MinecraftServer s, String type, Registry<T> registry) {
        this.server = s;
        this.registry = registry;
        this.registryName = (ResourceKey<Registry<T>>) registry.key();
        this.type = type;
    }

    public void restrict(String resourceName, @NotNull Consumer<B> consumer) {
        RegistrationType<T> registrationType = new RegistrationType<T>(resourceName, registryName);

        registrationType.ifNamespace(namespace -> restrictNamespace(namespace, consumer));
        registrationType.ifName(name -> restrictOne(name, consumer));
        registrationType.ifTag(tag -> restrictTag(tag, consumer));
    }

    @HideFromJS
    abstract protected void restrictOne(ResourceLocation name, @NotNull Consumer<B> consumer);

    @HideFromJS
    protected void logRestrictionCreation(R restriction, ResourceLocation name) {
        var inBiomes = appendListInfo(restriction.includeBiomes, "in biomes");
        var notInBiomes = appendListInfo(restriction.excludeBiomes, "not in biomes");
        var inDimensions = appendListInfo(restriction.includeDimensions, "in dimensions");
        var notInDimensions = appendListInfo(restriction.excludeDimensions, "not in dimensions");
        var details = Stream.of(inBiomes, notInBiomes, inDimensions, notInDimensions)
                .filter(value -> value.length() > 0)
                .collect(Collectors.joining(", "));

        ConsoleJS.SERVER.info("Created " + type + " restriction for " + name + " " + details);
    }

    @HideFromJS
    private String appendListInfo(List<ResourceLocation> list, String description) {
        return list.isEmpty() ? "" : description + " " + list;
    }

    @HideFromJS
    private void restrictNamespace(String namespace, @NotNull Consumer<B> consumer) {
        ConsoleJS.SERVER.info("Creating " + type + " restrictions for namespace " + namespace);

        registry.keySet()
                .stream()
                .filter(name -> name.getNamespace().equals(namespace))
                .forEach(name -> restrictOne(name, consumer));
    }

    @HideFromJS
    abstract public Predicate<T> isTagged(TagKey<T> tag);

    @HideFromJS
    abstract public ResourceLocation getName(T resource);

    @HideFromJS
    private void restrictTag(TagKey<T> tag, @NotNull Consumer<B> consumer) {
        ConsoleJS.SERVER.info("Creating " + type + " restrictions for tag " + tag.location());

        registry.stream()
                .filter(isTagged(tag))
                .forEach(fluid -> restrictOne(getName(fluid), consumer));
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/kubejs/api/AbstractRegistrationEventJS.kt
package net.impleri.playerskills.integrations.kubejs.api

import dev.latvian.mods.kubejs.server.ServerEventJS
import dev.latvian.mods.kubejs.util.ConsoleJS
import dev.latvian.mods.rhino.util.HideFromJS
import net.impleri.playerskills.restrictions.AbstractRestriction
import net.impleri.playerskills.utils.RegistrationType
import net.minecraft.core.Registry
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraft.tags.TagKey

abstract class AbstractRegistrationEventJS<T, R : AbstractRestriction<T>, B : AbstractRestrictionBuilder<R>>(
  s: MinecraftServer,
  private val type: String,
  private val registry: Registry<T>,
) : ServerEventJS(s) {
  private val registryName = registry.key() as ResourceKey<Registry<T>>

  fun restrict(resourceName: String, consumer: (B) -> Unit) {
    val registrationType = RegistrationType(resourceName, registryName)
    registrationType.ifNamespace { restrictNamespace(it, consumer) }
    registrationType.ifName { restrictOne(it, consumer) }
    registrationType.ifTag { restrictTag(it, consumer) }
  }

  @HideFromJS
  protected abstract fun restrictOne(name: ResourceLocation, consumer: (B) -> Unit)

  @HideFromJS
  protected fun logRestrictionCreation(restriction: R, name: ResourceLocation) {
    val inBiomes = appendListInfo(restriction.includeBiomes, "in biomes")
    val notInBiomes = appendListInfo(restriction.excludeBiomes, "not in biomes")
    val inDimensions = appendListInfo(restriction.includeDimensions, "in dimensions")
    val notInDimensions = appendListInfo(restriction.excludeDimensions, "not in dimensions")

    val details = listOf(inBiomes, notInBiomes, inDimensions, notInDimensions)
      .filter { it.isNotEmpty() }
      .joinToString(", ")

    ConsoleJS.SERVER.info("Created $type restriction for $name $details")
  }

  @HideFromJS
  private fun appendListInfo(list: List<ResourceLocation>, description: String): String {
    return if (list.isEmpty()) "" else "$description $list"
  }

  @HideFromJS
  private fun restrictNamespace(namespace: String, consumer: (B) -> Unit) {
    ConsoleJS.SERVER.info("Creating $type restrictions for namespace $namespace")

    registry.keySet()
      .filter { it.namespace == namespace }
      .forEach { restrictOne(it, consumer) }
  }

  @HideFromJS
  abstract fun isTagged(tag: TagKey<T>): (T) -> Boolean

  @HideFromJS
  abstract fun getName(resource: T): ResourceLocation

  @HideFromJS
  private fun restrictTag(tag: TagKey<T>, consumer: (B) -> Unit) {
    ConsoleJS.SERVER.info("Creating $type restrictions for tag ${tag.location}")
    registry
      .filter(isTagged(tag))
      .forEach { restrictOne(getName(it), consumer) }
  }
}
